src/rules/button_has_type.rs

```diff
@@ -0,0 +1,362 @@
+  fn tags(&self) -> &'static [&'static str] {
+    &["react", "jsx", "fresh"]
+  }
+
+  fn code(&self) -> &'static str {
+    CODE
+  }
+
+  fn lint_program_with_ast_view(
+    &self,
+    context: &mut Context,
+    program: Program,
+  ) {
+    HasButtonTypeHandler.traverse(program, context);
+  }
+
+  #[cfg(feature = "docs")]
+  fn docs(&self) -> &'static str {
+    include_str!("../../docs/rules/button_has_type.md")
+  }
+}
+
+enum DiagnosticKind {
+  MissingTypeAttr,
+  MissingValue,
+  WrongValue,
+}
+
+impl DiagnosticKind {
+  #[cfg(test)]
+  fn message_and_hint(&self) -> (&'static str, &'static str) {
+    (self.message(), self.hint())
+  }
+
+  fn message(&self) -> &'static str {
+    match *self {
+      DiagnosticKind::MissingTypeAttr => {
+        "`button` elements must have a `type` attribute"
+      }
+      DiagnosticKind::MissingValue => {
+        "Missing value for button `type` attribute"
+      }
+      DiagnosticKind::WrongValue => {
+        "Incorrect value for button `type` attribute"
+      }
+    }
+  }
+
+  fn hint(&self) -> &'static str {
+    match *self {
+      DiagnosticKind::MissingTypeAttr => {
+        "Add a `type` attribute with a value of `button`, `submit` or `reset`"
+      }
+      DiagnosticKind::MissingValue => "The value of the `type` attribute must be one of `button`, `submit` or `reset`",
+      DiagnosticKind::WrongValue => "The value of the `type` attribute must be one of `button`, `submit` or `reset`",
+    }
+  }
+}
+
+struct HasButtonTypeHandler;
+
+impl Handler for HasButtonTypeHandler {
+  fn jsx_opening_element(
+    &mut self,
+    node: &ast_view::JSXOpeningElement,
+    ctx: &mut Context,
+  ) {
+    if let JSXElementName::Ident(id) = node.name {
+      if id.sym() == "button" {
+        let mut found = false;
+        for attr in node.attrs {
+          if let JSXAttrOrSpread::JSXAttr(attr) = attr {
+            if let JSXAttrName::Ident(name) = attr.name {
+              if name.sym() == "type" {
+                found = true;
+
+                if let Some(attr_value) = attr.value {
+                  let kind = DiagnosticKind::WrongValue;
+                  match attr_value {
+                    JSXAttrValue::Lit(lit) => {
+                      if let Lit::Str(lit_str) = lit {
+                        let value = lit_str.value();
+                        if !is_valid_value(value) {
+                          ctx.add_diagnostic_with_hint(
+                            attr_value.range(),
+                            CODE,
+                            kind.message(),
+                            kind.hint(),
+                          );
+                        }
+                      } else {
+                        ctx.add_diagnostic_with_hint(
+                          attr_value.range(),
+                          CODE,
+                          kind.message(),
+                          kind.hint(),
+                        );
+                      }
+                    }
```

Comment:
Ouch, the indentation here 😬 can you please use `let ... else ...` or the `if-chain` crate that we already depend on?